A couple scripts for wheel building.

exarkun opened the pull request:
Fixes https://tahoe-lafs.org/trac/pycryptopp/ticket/106

This doesn't integrate with CI at all because travis wants AWS credentials so it can upload build artifacts to S3 and at the moment I don't have any such credentials to supply.  It seems like it might not be unreasonable to merely have the release engineer run this script as part of the release process.

I can look into CI more if that's particularly desired, though.

exarkun committed:
A couple scripts for wheel building.

exarkun committed:
Only build Python 2.x wheels.

exarkun committed:
auditwheel is not applicable to pycryptopp and it seems to screw up

exarkun committed:
Use a wheelhouse we can see outside of the container.

exarkun commented:
After merging, wheels for the last release could be generated now with:

```
git checkout pycryptopp-0.7.1
git checkout origin/master -- misc/build_helpers/_build-wheels.sh misc/build_helpers/build-manylinux-wheels.sh
misc/build_helpers/build-manylinux-wheels.sh
```

And uploaded to PyPI.

exarkun committed:
Maybe auditwheel does do something, I don't know.

david415 commented:
cool! this looks good.

having the release manager build the wheel files with a script sounds fine. having a remote SPOF like travis do it does not sounds so good. having both local and remote SPOFs do it and then compare hashes sounds even better ;-p

david415 commented on misc/build_helpers/_build-wheels.sh:
```diff
@@ -0,0 +1,16 @@
+
+# Based on https://github.com/pypa/python-manylinux-demo
+
+set -e -x
```
heh yeah

david415 commented on misc/build_helpers/_build-wheels.sh:
```diff
@@ -0,0 +1,16 @@
+set -e -x
+
+# Compile wheels
+for PYBIN in /opt/python/cp2*/bin; do
```
/opt/python looks like a mistake... is that what the s4 containers have?

exarkun replied:
That's what the Docker images being used have.  I don't know why they picked that path in particular but the images include a number of different Python builds (each ABI incompatible with the rest).  The output of the build is one wheel per Python ABI.

david415 commented on misc/build_helpers/_build-wheels.sh:
```diff
@@ -0,0 +1,16 @@
+
+# Bundle external shared libraries into the wheels and write them to their
+# final location.
+for whl in /wheelhouse/pycrypto-*.whl; do
```
must we have a directory in root called wheelhouse or can we have a build directory somewhere reasonable ;-p

david415 replied:
oh or is this meant to run in the docker image?

exarkun replied:
Yea this is only meant to run inside Docker.

exarkun committed:
Get the file name right

david415 commented on misc/build_helpers/build-manylinux-wheels.sh:
```diff
@@ -0,0 +1,17 @@
+BUILDER="/io/misc/build_helpers/_build-wheels.sh"
+
+# This image can build x86_64 (64 bit) manylinux wheels.
+DOCKER_IMAGE="quay.io/pypa/manylinux1_x86_64"
```
hmm a docker image from quay.io? it's too bad there isn't a threshold signature scheme at work here to prevent us to installing a bad container image in the case where quay.io get hacked.

exarkun replied:
I'll see if there's a way we can use Docker's image signature feature ... which may or may not help with confidence we're running what we want to run.

exarkun replied:
https://docs.docker.com/engine/security/trust/content_trust/ describes the system.  It seems to work with quay.io.  So we could embed the sha256 of the expect image into the script and, if we trust Docker's "content trust" system then that ensures we're using exactly the image we wanted to.

Getting the sha256 of the image we want to use is a little trickier.  quay.io builds the image so even if I audit the Dockerfile I can't really know the image quay.io is serving me _now_ actually agrees with that Dockerfile.  Maybe I can build the image myself, locally, and it will end up with the same sha256?  But I doubt it because timestamps and stuff.

exarkun replied:
Alternatively, we could suggest building the Docker images locally from the Dockerfiles on github - eg https://github.com/pypa/manylinux/blob/master/docker/Dockerfile-x86_64

david415 reviewed (commented):
looks good but does this mean the release engineer must use docker?

exarkun commented:
> looks good but does this mean the release engineer must use docker?

The Docker images being used are based on very old Linux distributions so that the libraries that get linked against are very old.  The compatibility guarantees you get from these builds is that the wheels will work on a system that has versions of the libraries linked against at least as new as the ones in the build environment.  Technically it's possible to do a similar build without Docker but it involves getting very old libraries into your build environment somehow.  That's a lot more work.  So, sad as a Docker dependency is, it's probably the least bad way.

exarkun committed:
Fetch dependencies outside the container

Disable networking for the container.

The container images run very old, vulnerability-filled software.
(The former is intentional, the latter is somewhat difficult to avoid.)

This allows the containers to do the wheel builds without talking to the
network.  This reduces the attack surface somewhat.

warner commented:
looks good, merging
